### PR TITLE
[oneDNN]: Added support for fp16 for oneDNN ops

### PR DESCRIPTION
**Pull request by bhavani-subramanian**
This PR adds FP16 support for the following ops: convolution, matmul (mkl-matmul, batch-matmul, fused-matmul), maxpool, batchnorm, layernorm and softmax.

This PR also adds FP16 ISA guards in C++ and Python so that oneDNN related fusions in remapper and node-rewrites in layout pass happen only on platforms with AVX-512 support. Further, BF16 ISA guards in C++ are refactored into a common function that accepts `DT_FLOAT`, `DT_BFLOAT16` and `DT_HALF` inputs. It can be easily extended in the future for new types as well (ex. `float8`).

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 1]
```diff
@@ -1946,13 +1948,8 @@ class RemapperFuseMatMulWithBiasAndActivationTest : public RemapperTest {
...
   void RunTest() {
     using ::tensorflow::ops::Placeholder;
-
-#if defined(INTEL_MKL) && defined(ENABLE_MKL)
-    std::vector<string> activations = {"Relu", "Relu6", "Elu", "Tanh",
-                                       "LeakyRelu"};
-#else
     std::vector<string> activations = {"Relu", "Relu6", "Elu", "LeakyRelu"};
-#endif
+    if (IsMKLEnabled()) activations.push_back("Tanh");
```
Comment (cantonios) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:1952]: Are you sure it's only enabled with MKL?  I believe I added this for CPU in commit 8cb3e0a4901bb7d5f62acc033220408ade3dc4e8.

File: tensorflow/core/ops/math_ops.cc [hunk 2]
```diff
@@ -167,7 +167,7 @@ REGISTER_OP("_MklBatchMatMul")
     .Input("x: T")
```
Comment (cantonios) [inline tensorflow/core/ops/math_ops.cc:167]: Maybe these ops should be moved to a separate file, like the was done for nn ops?

**Review by cantonios — APPROVED** [review]
Minor comments, approving to trigger tests

File: tensorflow/core/grappler/optimizers/remapper_test.cc (hunk 1 above)
Comment (bhavani-subramanian) [reply tensorflow/core/grappler/optimizers/remapper_test.cc]: Good catch! Fixed it.

File: tensorflow/core/ops/math_ops.cc (hunk 2 above)
Comment (bhavani-subramanian) [reply tensorflow/core/ops/math_ops.cc:170]: Is it okay if I do it in a separate PR? Because it would require more changes and additional testing as well. Please let me know. Thanks.

**Comment by bhavani-subramanian** [comment]
@cantonios Thanks for reviewing this PR. I have addressed your review comments. Please take a look. Thanks.

File: tensorflow/core/grappler/optimizers/remapper_test.cc (hunk 1 above)
Comment (bhavani-subramanian) [reply tensorflow/core/grappler/optimizers/remapper_test.cc]: Looks like removing the MKL check is causing the test to fail on GPU. I've fixed it by disabling the test on GPU.

**Review by cantonios — APPROVED** [review]

File: tensorflow/core/kernels/mkl/mkl_tmp_ops.cc [hunk 3]
```diff
@@ -21,34 +21,48 @@ limitations under the License.
...
-class RaiseBfloat16Error : public OpKernel {
+template <typename T>
+class RaiseError : public OpKernel {
  public:
-  explicit RaiseBfloat16Error(OpKernelConstruction* context)
-      : OpKernel(context) {
+  explicit RaiseError(OpKernelConstruction* context) : OpKernel(context) {
     OP_REQUIRES(context, false,
-                errors::InvalidArgument("Op does not support bfloat16 inputs"));
+                errors::InvalidArgument(absl::StrCat(
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_tmp_ops.cc:36]: Please use `absl::InvalidArgumentError` instead.
```suggestion
                absl::InvalidArgumentError(absl::StrCat(
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Sorry for the delay! Feedback from internal code check.

File: tensorflow/core/kernels/mkl/mkl_tmp_ops.cc (hunk 3 above)
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_tmp_ops.cc]: Fixed here and in other files modified by this PR.

**Review by penpornk — APPROVED** [review]
Thank you for the quick fixes!

**Comment by bhavani-subramanian** [comment]
@penpornk Thanks for approving this PR!  Looks like `feedback/copybara` checks have failed. Can you please share the test failures since I am unable to access it? Thanks.
